subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformParallelIntegrationTest.groovy

```diff
@@ -410,4 +417,51 @@ class ArtifactTransformParallelIntegrationTest extends AbstractDependencyResolut
         then:
         handle.waitForFinish()
     }
+
+    def "only one process can run immutable transforms at the same time"() {
+        given:
+        def buildA = new BuildTestFile(file("buildA"), "buildA")
+        def buildB = new BuildTestFile(file("buildB"), "buildB")
+        def toBeTransformed = file("lib.jar")
+        toBeTransformed.text = '1234'
+
+        [buildA, buildB].each { build ->
+            setupBuild(build)
+            build.with {
+                def a = file(build.rootProjectName + ".jar")
+                a.text = '1234'
+                buildFile << """
+                    dependencies {
+                        compile files(name + ".jar")
+                    }
+                    task resolve {
+                        def artifacts = configurations.compile.incoming.artifactView {
+                            attributes { it.attribute(artifactType, 'size') }
+                        }.artifacts
+                        inputs.files(artifacts.artifactFiles)
+
+                        doLast { 
+                            ${server.callFromBuildUsingExpression('"resolveStarted_" + project.name')}
+                            assert artifacts.artifactFiles.collect { it.name } == [project.name + '.jar.txt']
+                        }
+                    }
+                """
+            }
+        }
+
+        expect:
+        server.expectConcurrent("resolveStarted_buildA", "resolveStarted_buildB")
+        def transformations = server.expectConcurrentAndBlock(1, "buildA.jar", "buildB.jar")
+        def buildAHandle = executer.inDirectory(buildA).withTasks("resolve").start()
+        def buildBHandle = executer.inDirectory(buildB).withTasks("resolve").start()
+
+        transformations.waitForAllPendingCalls()
+        Thread.sleep(1000)
```

Comment:
Why sleep here?

Reply:
`waitForAllPendingCalls` waits until one transform starts being executed. If we don't wait here, we don't give the second transform a chance to execute in parallel.

Without the locking, the test didn't fail without the sleep and started failing with the sleep if I don't do any locking. Am I doing something wrong here?

Reply:
I changed the test a little bit, though I wasn't able to get rid of the `Thread.sleep` :(
Please take another look.